storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java

```diff
@@ -709,6 +720,33 @@ public void assign(SchedulerAssignment assignment, boolean ignoreSingleException
         }
     }
 
+    /**
+     * Initialize the flag to true if specified topology uses SharedOffHeapNodeMemory, false otherwise.
+     *
+     * @param td TopologyDetails to examine
+     */
+    private void initializeTopoSharedOffHeapNodeMemoryFlag(TopologyDetails td) {
+        String topoId = td.getId();
+        topoSharedOffHeapMemoryNodeFlag.put(topoId, false);
+        StormTopology topology = td.getTopology();
+        if (topology == null) {
+            return; // accomodate multitenant_scheduler_test.clj
+        }
+        if (topology.is_set_component_to_shared_memory()) {
```

Comment:
We should be able to simplify this with 

```
        StormTopology topology = td.getTopology();
        if (topology.is_set_shared_memory()) {
            for (SharedMemory sharedMemory : topology.get_shared_memory().values()) {
                double val = sharedMemory.get_off_heap_node();
                if (val > 0.0) {
                    topoSharedOffHeapMemoryNodeFlag.put(topoId, true);
                    return;
                }
            }
        }
        
```

Reply:
(1) topology == null check is required weirdly enough due to some other tests failure (later I also found note in TopologyDetails.getSharedMemoryRequests()
(2) In StormTopology shared_memory and component_to_shared_memory are two different thrift fields. And setters don't seem to cross set. If I change this code then this will change the prior logic in Cluster.java. How are these two different fields equivalent?

Reply:
I agree with (1). Sorry I didn't copy that part of the code. 

(2) 
These two fields are equivalent. 
There is a 1-to-1 mapping between `SharedMemory name` to `SharedMemory object`.

shared_memory and component_to_shared_memory are different structures to keep `SharedMemory object` but they essentially keep the whole mapping of `SharedMemory name` to `SharedMemory object`
https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/topology/TopologyBuilder.java#L631-L638

Reply:
They are designed this way. For example

I can have two SharedMemory with name as `sharedMemory1` and `sharedMemory2`. 

And I can add  `sharedMemory1` to `comp1` and `sharedMemory2` to `comp2`.

Now shared_memory is 
 object1 (name=`sharedMemory1`),  object2 (name=`sharedMemory2`);

component_to_shared_memory is 

`comp1` -> object1 (name=`sharedMemory1`), `comp2` -> object2 (name=`sharedMemory2`);

Reply:
Thanks for explanation. Looks like TopologyBuilder.createTopology() makes the only to call to set_component_to_shared_memory() and set_shared_memory() and thus ensuring this consistency.

Changed code.